app/src/main/java/com/keylesspalace/tusky/util/SpanUtils.kt

```diff
@@ -0,0 +1,132 @@
+package com.keylesspalace.tusky.util
+
+import android.text.Spannable
+import android.text.Spanned
+import android.text.style.CharacterStyle
+import android.text.style.ForegroundColorSpan
+import android.text.style.URLSpan
+import java.util.regex.Pattern
+
+class SpanUtils {
+    companion object {
+        /**
+         * @see <a href="https://github.com/tootsuite/mastodon/blob/master/app/models/tag.rb">
+         *     Tag#HASHTAG_RE</a>.
+         */
+        private const val TAG_REGEX = "(?:^|[^/)\\w])#([\\w_]*[\\p{Alpha}_][\\w_]*)"
+
+        /**
+         * @see <a href="https://github.com/tootsuite/mastodon/blob/master/app/models/account.rb">
+         *     Account#MENTION_RE</a>
+         */
+        private const val MENTION_REGEX = "(?:^|[^/[:word:]])@([a-z0-9_]+(?:@[a-z0-9\\.\\-]+[a-z0-9]+)?)"
+
+        private const val HTTP_URL_REGEX = "(?:(^|\\b)http://[^\\s]+)"
+        private const val HTTPS_URL_REGEX = "(?:(^|\\b)https://[^\\s]+)"
+
+        private val spanClasses = listOf(ForegroundColorSpan::class.java, URLSpan::class.java)
+        private val finders = mapOf(
+            FoundMatchType.HTTP_URL to PatternFinder(':', HTTP_URL_REGEX, 5),
+            FoundMatchType.HTTPS_URL to PatternFinder(':', HTTPS_URL_REGEX, 6),
+            FoundMatchType.TAG to PatternFinder('#', TAG_REGEX, 1),
+            FoundMatchType.MENTION to PatternFinder('@', MENTION_REGEX, 1)
+        )
+
+        private enum class FoundMatchType {
+            HTTP_URL,
+            HTTPS_URL,
+            TAG,
+            MENTION,
+        }
+
+        private class FindCharsResult {
+            lateinit var matchType: FoundMatchType
```

Comment:
Why latenit? This could be even a data class I guess

Reply:
Okay, it is used in imperative code which is imperative for the sake of performance I guess but I would avoid lateinit anyway, if it's optional - let it be so